src/crypto/generichash/mod.rs

```diff
@@ -0,0 +1,211 @@
+//! `GenericHash`.
+//!
+use ffi::{
+    crypto_generichash_BYTES_MAX, crypto_generichash_BYTES_MIN,
+    crypto_generichash_KEYBYTES_MAX, crypto_generichash_KEYBYTES_MIN,
+    crypto_generichash_final, crypto_generichash_init,
+    crypto_generichash_state, crypto_generichash_update,
+};
+
+use libc::c_ulonglong;
+use std::mem;
+use std::ptr;
+
+mod digest;
+pub use self::digest::Digest;
+
+/// Minimium of allowed bytes in a `Digest`
+pub const DIGEST_MIN: usize = crypto_generichash_BYTES_MIN as usize;
+
+/// Maximum of allowed bytes in a `Digest`
+pub const DIGEST_MAX: usize = crypto_generichash_BYTES_MAX as usize;
+
+/// Minimium of allowed bytes in a key
+pub const KEY_MIN: usize = crypto_generichash_KEYBYTES_MIN as usize;
+
+/// Maximum of allowed bytes in a key
+pub const KEY_MAX: usize = crypto_generichash_KEYBYTES_MAX as usize;
+
+/// `State` contains the state for multi-part (streaming) hash computations. This allows the caller
+/// to process a message as a sequence of multiple chunks.
+pub struct State {
+    out_len: usize,
+    state: crypto_generichash_state,
+}
+
+impl State {
+    /// `new` constructs and initializes a new `State` with the given parameters.
+    ///
+    /// `out_len` specifies the resulting hash size.
+    /// Only values in the interval [`DIGEST_MIN`, `DIGEST_MAX`] are allowed.
+    ///
+    /// `key` is an optional parameter, which when given,
+    /// a custom key can be used for the computation of the hash.
+    /// The size of the key must be in the interval [`KEY_MIN`, `KEY_MAX`].
+    pub fn new(out_len: usize, key: Option<&[u8]>) -> Result<State, ()> {
+        if out_len < DIGEST_MIN || out_len > DIGEST_MAX {
+            return Err(());
+        }
+
+        if let Some(key) = key {
+            let len = key.len();
+            if len < KEY_MIN || len > KEY_MAX {
+                return Err(());
+            }
+        }
+
+        let mut state: crypto_generichash_state;
+
+        let result = unsafe {
+            state = mem::uninitialized();
+            if let Some(key) = key {
+                crypto_generichash_init(
+                    &mut state,
+                    key.as_ptr(),
+                    key.len(),
+                    out_len,
+                )
+            } else {
+                crypto_generichash_init(&mut state, ptr::null(), 0, out_len)
+            }
+        };
+
+        if result == 0 {
+            Ok(State {
+                out_len: out_len,
+                state,
+            })
+        } else {
+            Err(())
+        }
+    }
+
+    /// `update` updates the `State` with `data`. `update` can be called multiple times in order
+    /// to compute the hash from sequential chunks of the message.
+    pub fn update(&mut self, data: &[u8]) -> Result<(), ()> {
+        if unsafe {
```

Comment:
Since you fixed the other one, note this is still here.

Reply:
XD.